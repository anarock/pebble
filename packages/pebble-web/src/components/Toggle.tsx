import * as React from "react";
import { cx } from "emotion";
import {
  labelStyle,
  selectedLabel,
  inputStyle,
  selectedSwitch,
  switchStyle,
  fixedLabelStyle
} from "./styles/Toggle.styles";

interface ToggleProps {
  initialValue?: boolean;
  onChange?: (value: boolean) => void;
  className?: string;
}

interface ToggleState {
  value: boolean;
}

export default class Toggle extends React.PureComponent<
  ToggleProps,
  ToggleState
> {
  state = {
    value: this.props.initialValue || false
  };

  render() {
    const { onChange, className } = this.props;
    const { value } = this.state;

    return (
      <label
        className={cx(labelStyle, className, fixedLabelStyle, {
          [selectedLabel]: value
        })}
      >
        <input
          type="checkbox"
          onChange={e => {
            this.setState({
              value: e.target.checked
            });
            if (onChange) {
              onChange(e.target.checked);
            }
          }}
          checked={value}
          className={inputStyle}
        />
        <span
          className={cx({
            [switchStyle]: true,
            [selectedSwitch]: value
          })}
        />
      </label>
    );
  }
}
