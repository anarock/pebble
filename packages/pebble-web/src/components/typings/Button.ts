import * as React from "react";
import * as CSS from "csstype";
import { Interpolation } from "./emotionCustom";

export type ButtonType = "primary" | "secondary" | "link" | "success" | "alert";

export interface ButtonProps {
  children: React.ReactNode;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  showShadow?: boolean;
  disabled?: boolean;
  type?: ButtonType;
  size?: "x-small" | "small" | "large";
  width?: CSS.WidthProperty<number>;
  styles?: Interpolation;
  showRipple?: boolean;
  loading?: boolean;
  filled?: boolean;
  buttonProps?: React.ButtonHTMLAttributes<HTMLButtonElement>;
}

export interface DropDownButtonProps extends ButtonProps {
  isOpen: boolean;
  isSelected: boolean;
}

interface MappingColorByTypeOptions {
  textColor?: string;
  base: string;
  hover: string;
  active: string;
  disabled: string;
}

export interface MappingColorByType {
  primary: MappingColorByTypeOptions;
  secondary: MappingColorByTypeOptions;
  success: MappingColorByTypeOptions;
  alert: MappingColorByTypeOptions;
}
