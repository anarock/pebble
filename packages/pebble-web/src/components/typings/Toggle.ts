export interface ToggleProps {
  value: "ON" | "OFF";
  onButtonText: string;
  offButtonText: string;
  onChange: () => void;
}
