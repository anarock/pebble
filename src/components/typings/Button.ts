import * as React from "react";

export type ButtonType = "primary" | "secondary" | "link" | "success" | "alert";

type Children = JSX.Element | string | number | React.ReactChildren;

export interface ButtonProps {
  children: Children;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  showShadow?: boolean;
  disabled?: boolean;
  type?: ButtonType;
  size?: "x-small" | "small" | "large";
  width?: number;
  className?: string;
  showRipple?: boolean;
  loading?: boolean;
  filled?: boolean;
}

export interface DropDownButtonProps extends ButtonProps {
  isOpen: boolean;
  isSelected: boolean;
}

interface MappingColorByTypeOptions {
  textColor?: string;
  base: string;
  hover: string;
  active: string;
  disabled: string;
}

export interface MappingColorByType {
  primary: MappingColorByTypeOptions;
  secondary: MappingColorByTypeOptions;
  success: MappingColorByTypeOptions;
  alert: MappingColorByTypeOptions;
}
