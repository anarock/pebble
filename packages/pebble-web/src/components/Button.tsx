import * as React from "react";

import {
  iconStyle,
  getButtonStyle,
  dropDownButtonStyle,
  dropDownButtonDefaultStyle
} from "./styles/Button.styles";
import Ink from "react-ink";
import { ButtonProps, DropDownButtonProps } from "./typings/Button";
import Loader from "./Loader";
import { colors } from "pebble-shared";

const Button: React.FunctionComponent<ButtonProps> = ({
  type = "primary",
  disabled,
  children,
  onClick,
  width,
  showShadow,
  styles,
  showRipple = true,
  loading,
  size = "small",
  buttonProps
}: ButtonProps) => {
  const disableAction = disabled || loading;

  const filled = size !== "x-small";

  return (
    <button
      css={[getButtonStyle(size, type, !!showShadow, filled), styles]}
      onClick={!disableAction ? onClick : undefined}
      style={{ width }}
      disabled={disabled}
      {...buttonProps}
    >
      {loading ? <Loader color={colors.white.base} scale={0.4} /> : children}
      {!disableAction && showRipple && type !== "link" && <Ink />}
    </button>
  );
};

export const DropDownButton = ({
  isOpen,
  isSelected,
  children,
  styles,
  ...props
}: DropDownButtonProps) => {
  return (
    <Button
      {...props}
      type="secondary"
      styles={[
        dropDownButtonStyle,
        !(isOpen || isSelected) && dropDownButtonDefaultStyle,
        styles
      ]}
    >
      {children}{" "}
      <i
        className="pi pi-arrow-drop-down"
        css={iconStyle}
        style={{
          transform: isOpen ? "rotate(180deg)" : "none"
        }}
      />
    </Button>
  );
};

export default Button;
