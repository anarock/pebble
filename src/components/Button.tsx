import * as React from "react";
import { cx } from "emotion";
import {
  iconStyle,
  getButtonStyle,
  dropDownButtonStyle,
  dropDownButtonDefaultStyle
} from "./styles/Button.styles";
import Ink from "react-ink";
import { ButtonProps, DropDownButtonProps } from "./typings/Button";
import Loader from "./Loader";
import { colors } from "../theme";

const Button: React.FunctionComponent<ButtonProps> = ({
  type = "primary",
  disabled,
  children,
  onClick,
  width,
  showShadow,
  className,
  showRipple = true,
  loading,
  size = "small"
}: ButtonProps) => {
  const disableAction = disabled || loading;

  const filled = size !== "x-small";
  const _className = cx(
    getButtonStyle(size, type, !!showShadow, filled),
    className
  );

  return (
    <button
      className={_className}
      onClick={!disableAction ? onClick : undefined}
      style={{ width }}
      disabled={disabled}
    >
      {loading ? <Loader color={colors.white.base} scale={0.4} /> : children}
      {!disableAction && showRipple && type !== "link" && <Ink />}
    </button>
  );
};

export const DropDownButton: React.FunctionComponent<DropDownButtonProps> = ({
  isOpen,
  isSelected,
  children,
  className,
  ...props
}) => {
  const _className = cx(dropDownButtonStyle, {
    [dropDownButtonDefaultStyle]: !(isOpen || isSelected)
  });

  return (
    <Button {...props} type="secondary" className={cx(_className, className)}>
      <React.Fragment>
        {children}{" "}
        <i
          className={cx("pi pi-arrow-drop-down", iconStyle)}
          style={{
            transform: isOpen ? "rotate(180deg)" : "none"
          }}
        />
      </React.Fragment>
    </Button>
  );
};

export default Button;
