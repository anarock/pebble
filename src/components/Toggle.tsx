import * as React from "react";
import Button from "./Button";
import { cx } from "emotion";
import {
  toggleWrap,
  offButton,
  selectedButton,
  onButton
} from "./styles/Toggle.style";
import { ToggleProps } from "./typings/Toggle";

const Toggle: React.SFC<ToggleProps> = props => {
  const { value, onButtonText, offButtonText, onChange } = props;
  return (
    <div className={toggleWrap}>
      <Button
        onClick={onChange}
        type="secondary"
        size="x-small"
        className={cx({
          [offButton]: true,
          [selectedButton]: value === "OFF"
        })}
        showRipple={false}
      >
        {onButtonText}
      </Button>
      <Button
        onClick={onChange}
        type="secondary"
        size="x-small"
        className={cx({
          [onButton]: true,
          [selectedButton]: value === "ON"
        })}
        showRipple={false}
      >
        {offButtonText}
      </Button>
    </div>
  );
};

export default Toggle;
