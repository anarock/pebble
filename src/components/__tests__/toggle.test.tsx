import renderer from "react-test-renderer";
import combos from "combos";
import Toggle from "../Toggle";
import * as React from "react";
import { mount } from "enzyme";
import sinon from "sinon";
import Button from "../Button";

function noop() {}

describe("Toggle", () => {
  const _props = {
    value: ["ON", "OFF"],
    onButtonText: ["AM"],
    offButtonText: ["PM"]
  };
  const _combos = combos(_props);

  test.each(_combos)("%o", props => {
    const timePicker = renderer.create(<Toggle {...props} onChange={noop} />);

    const tree = timePicker.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("TimePicker: functionality test", () => {
  test("should call onChange on click", () => {
    const spy = sinon.spy();

    const toggle = mount(
      <Toggle
        value={"ON"}
        onChange={spy}
        onButtonText="AM"
        offButtonText="PM"
      />
    );

    toggle
      .find(Button)
      .at(0)
      .simulate("click");

    expect(spy.calledOnce).toBeTruthy();
  });
});
