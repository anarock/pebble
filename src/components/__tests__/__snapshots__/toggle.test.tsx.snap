// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle { value: 'OFF', onButtonText: 'AM', offButtonText: 'PM' } 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  line-height: 23px;
  height: 40px;
  padding: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all ease-out .2s;
  transition: all ease-out .2s;
  border-radius: 3px;
  outline: none;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  white-space: nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  height: 25px;
  min-width: 70px;
  font-weight: 400;
  color: #EDEDED;
  font-size: 10px;
  background-color: #ffffff;
  border: 1px solid #EDEDED;
  box-shadow: none;
  border-radius: 3px 0px 0px 3px;
  border-right: none;
  color: #101721;
  padding: 8px 20px;
  height: 40px;
  font-size: 12px;
  background-color: #EDEDED;
}

.emotion-0[disabled] {
  cursor: not-allowed;
}

.emotion-0:not([disabled]):hover {
  color: #101721;
  background-color: #F7F7F9;
  border-color: #F7F7F9;
}

.emotion-0:not([disabled]):active {
  color: #101721;
  background-color: #9FA6B0;
  border-color: #9FA6B0;
}

.emotion-0[disabled] {
  color: #101721;
  background-color: #EDEDED;
  border-color: #EDEDED;
}

.emotion-1 {
  line-height: 23px;
  height: 40px;
  padding: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all ease-out .2s;
  transition: all ease-out .2s;
  border-radius: 3px;
  outline: none;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  white-space: nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  height: 25px;
  min-width: 70px;
  font-weight: 400;
  color: #EDEDED;
  font-size: 10px;
  background-color: #ffffff;
  border: 1px solid #EDEDED;
  box-shadow: none;
  border-left: 1px solid #EDEDED;
  border-radius: 0px 3px 3px 0px;
  color: #101721;
  padding: 8px 20px;
  height: 40px;
  font-size: 12px;
}

.emotion-1[disabled] {
  cursor: not-allowed;
}

.emotion-1:not([disabled]):hover {
  color: #101721;
  background-color: #F7F7F9;
  border-color: #F7F7F9;
}

.emotion-1:not([disabled]):active {
  color: #101721;
  background-color: #9FA6B0;
  border-color: #9FA6B0;
}

.emotion-1[disabled] {
  color: #101721;
  background-color: #EDEDED;
  border-color: #EDEDED;
}

<div
  className="emotion-2"
>
  <button
    className="emotion-0"
    onClick={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
  >
    AM
  </button>
  <button
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
  >
    PM
  </button>
</div>
`;

exports[`Toggle { value: 'ON', onButtonText: 'AM', offButtonText: 'PM' } 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  line-height: 23px;
  height: 40px;
  padding: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all ease-out .2s;
  transition: all ease-out .2s;
  border-radius: 3px;
  outline: none;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  white-space: nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  height: 25px;
  min-width: 70px;
  font-weight: 400;
  color: #EDEDED;
  font-size: 10px;
  background-color: #ffffff;
  border: 1px solid #EDEDED;
  box-shadow: none;
  border-radius: 3px 0px 0px 3px;
  border-right: none;
  color: #101721;
  padding: 8px 20px;
  height: 40px;
  font-size: 12px;
}

.emotion-0[disabled] {
  cursor: not-allowed;
}

.emotion-0:not([disabled]):hover {
  color: #101721;
  background-color: #F7F7F9;
  border-color: #F7F7F9;
}

.emotion-0:not([disabled]):active {
  color: #101721;
  background-color: #9FA6B0;
  border-color: #9FA6B0;
}

.emotion-0[disabled] {
  color: #101721;
  background-color: #EDEDED;
  border-color: #EDEDED;
}

.emotion-1 {
  line-height: 23px;
  height: 40px;
  padding: 0 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all ease-out .2s;
  transition: all ease-out .2s;
  border-radius: 3px;
  outline: none;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  white-space: nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0;
  height: 25px;
  min-width: 70px;
  font-weight: 400;
  color: #EDEDED;
  font-size: 10px;
  background-color: #ffffff;
  border: 1px solid #EDEDED;
  box-shadow: none;
  border-left: 1px solid #EDEDED;
  border-radius: 0px 3px 3px 0px;
  color: #101721;
  padding: 8px 20px;
  height: 40px;
  font-size: 12px;
  background-color: #EDEDED;
}

.emotion-1[disabled] {
  cursor: not-allowed;
}

.emotion-1:not([disabled]):hover {
  color: #101721;
  background-color: #F7F7F9;
  border-color: #F7F7F9;
}

.emotion-1:not([disabled]):active {
  color: #101721;
  background-color: #9FA6B0;
  border-color: #9FA6B0;
}

.emotion-1[disabled] {
  color: #101721;
  background-color: #EDEDED;
  border-color: #EDEDED;
}

<div
  className="emotion-2"
>
  <button
    className="emotion-0"
    onClick={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
  >
    AM
  </button>
  <button
    className="emotion-1"
    onClick={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
  >
    PM
  </button>
</div>
`;
